pkgs/development/python-modules/pip/default.nix

```diff
@@ -14,40 +16,46 @@
 
 buildPythonPackage rec {
   pname = "pip";
-  version = "23.0.1";
-  format = "other";
+  version = "23.2.1";
+  format = "pyproject";
 
   src = fetchFromGitHub {
     owner = "pypa";
     repo = pname;
     rev = "refs/tags/${version}";
-    hash = "sha256-BSonlwKmegrlrQTTIL0avPi61/TY2M0f7kOZpSzPRQk=";
-    name = "${pname}-${version}-source";
+    hash = "sha256-mUlzfYmq1FE3X1/2o7sYJzMgwHRI4ib4EMhpg83VvrI=";
   };
 
-  nativeBuildInputs = [ bootstrapped-pip ];
-
   postPatch = ''
     # Remove vendored Windows PE binaries
     # Note: These are unused but make the package unreproducible.
     find -type f -name '*.exe' -delete
   '';
 
-  # pip detects that we already have bootstrapped_pip "installed", so we need
-  # to force it a little.
-  pipInstallFlags = [ "--ignore-installed" ];
+  nativeBuildInputs = [
+    installShellFiles
+    setuptools
+    wheel
+  ];
 
   nativeCheckInputs = [ mock scripttest virtualenv pretend pytest ];
+
   # Pip wants pytest, but tests are not distributed
   doCheck = false;
 
+  postInstall = ''
+    installShellCompletion --cmd pip \
+      --bash <($out/bin/pip completion --bash) \
+      --fish <($out/bin/pip completion --fish) \
+      --zsh <($out/bin/pip completion --zsh)
```

Comment:
`<($out/bin/pip completion --zsh --no-cache-dir)` will properly clear the warning about `.cache/pip` not being writable. zsh may not work until https://github.com/pypa/pip/issues/12166 is closed. (given pip is at the bottom of ~40000 packages leaving it in seems fine)

Reply:
Also fixes https://github.com/NixOS/nixpkgs/issues/224610

Reply:
Thank you for figuring that out. This was merged without that change, but we should add this in a separate commit.

Reply:
Please do. I'm working on another change atm and don't need the two things coupled.

Reply:
Tacked it on to this change https://github.com/NixOS/nixpkgs/pull/250398/files

Reply:
Thank you, I appreciate you creating that PR. I'm sorry I hadn't found the time yet today.